src/storage/txn/commands/prewrite.rs

```diff
@@ -528,6 +528,13 @@ impl<K: PrewriteKind> Prewriter<K> {
                     return check_committed_record_on_err(prewrite_result, txn, reader, &key);
                 }
                 Err(MvccError(box MvccErrorInner::CommitTsTooLarge { .. })) | Ok((..)) => {
+                    // The prewrite might be a retry and the record may have been committed.
+                    // So, we need to prevent the fallback to avoid duplicate commits.
+                    if let Ok(res) =
+                        check_committed_record_on_err(prewrite_result, txn, reader, &key)
```

Comment:
L530 matches `Ok((..))`. Is it expected? BTW, I forget why it matches `OK((..))` here.

Reply:
Sadly, this definitely causes some problems... If it needs min_commit_ts but min_commit_ts is zero, the lock
has been prewritten and has fallen back from async commit or 1PC.

Reply:
Yes.

Reply:
> I'm a bit confused. It seems `CommitTsTooLarge` is consumed inside and this error would not be returned no matter if it's a retried request or not?

`CommitTsTooLarge` is always consumed and not returned to the client...